packages/opentelemetry-api/src/trace/span_context.ts

```diff
@@ -67,4 +86,19 @@ export interface SpanContext {
    *         tracestate: rojo=00f067aa0ba902b7,congo=t61rcWkgMzE
    */
   traceState?: TraceState;
+
+  /**
+   * A function that determines whether the span context is valid
+   */
+  isValid(): boolean {
```

Comment:
I think I was following the other otel languages implementations. Most of them have `isValid` as part of the spanContext class. I can still just move the files if that's preferable, though.

Reply:
@obecny @mayurkale22 opinions on this? I don't really have a strong opinion either way. Both would be spec compliant in my opinion.

@xirzec do you have a strong opinion here, or were you just pointing out arguments?

To me, consistency with other implementations is a compelling argument.

Reply:
So let me try to capture my thoughts on this.

The main difficulty comes from the fact that JS lacks a type system. As such, it is not uncommon for one app to end up with two different versions of a package in its dependency graph, especially since NPM takes care of this for you and without a global type registry there are no conflicts.

However, it means that certain language constructs like `instanceof` don't always work as expected. If you try to check `if (context instanceof SpanContext)` it will only be true if the created object used the exact same constructor function as the one you provided. 

TypeScript tries to work around this, by comparing the public shape of classes, even if they come from different versions of a package, but it will fall down if there is any private differences in their shapes. E.g. if version 2 of a class has all the same public members as version 1 but it adds an additional private field, it will no longer be compatible.

Because of this, TS usually works better if a library takes and emits interfaces rather than classes since then you have broader compatibility across versions. If you can guarantee that a caller will **only** ever pass class instances to the same version of the library they created that instance object from, you'll be fine. 

However, that feels like a really strong stance to take for OpenTelemetry to take. Do you expect every app author and framework that supports OT to always be on latest? What if some component my app depends on doesn't rev fast enough? Am I forced to be locked on an old OT version until they do? What if two components snap to different versions?

I feel like this has been hand-waved away too much by automatic Span propagation, but any library that wants to support browsers (or older versions of NodeJS) is going to have to support manual Span propagation. And what is the easiest contract to do manual Span propagation right now? Passing a `SpanContext`. But now if it's a class rather than an interface, you're going to have to worry a lot more about compatibility breaks between versions.

Also by doing this change, every `SpanContext` object you had from a previous OT version will no longer be compatible with the new library and would force everyone to upgrade in a way that isn't backwards compatible. 

Maybe no other library author is using OT right now, but we've had enough pain in the Azure SDKs from OT breaking changes over the last year that I have to question the utility of continuing to try to support it. It seems like we'd be better off inventing our own solution that could export Spans through some kind of plugin model instead.

Reply:
Thank you for that very well reasoned argument. I agree that yes, that is quite compelling. Much more so than being similar to other implementations.

I am interested in the last paragraph:

>Maybe no other library author is using OT right now, but we've had enough pain in the Azure SDKs from OT breaking changes over the last year that I have to question the utility of continuing to try to support it. It seems like we'd be better off inventing our own solution that could export Spans through some kind of plugin model instead.

I'm very sorry you feel this way, and yes we do have very few third party libraries implementing right now. Certainly none as large as the Azure SDKs. While we have had quite a few breaking changes in the last year (the context changes being particularly painful on my side), I would certainly hope that we wouldn't push you to discontinue support altogether. Can you share which changes have been particularly painful to you?

Reply:
Sorry, I think my remarks came off as more extreme than what I intended. Allow me to quickly clarify and indeed we can continue off-thread if there needs to be further discussion.

As maintainers of this project are certainly familiar with, JS has no good native tracing story. In addition, the primitives required to do async callstack reconstruction aren't really exposed except to debuggers, with the notable exception of the work done in NodeJS by async hooks, which remains an experimental feature.

A lot of work was done last year to make Azure SDKs (across four languages: Python, C#, Java, and JS) reach "Generally Available" (GA) status in November 2019 and no longer accept breaking changes. One tricky part of that was we wanted to have a tracing story be part of that release. 

In languages that already had a tracing story (such as C#) we were able to leverage the existing platform support for tracing and create adapters that allowed for interop with OpenTelemetry (and in some legacy cases, OpenCensus.) This made it easy to react to breaking changes as the adapter itself need not be GA and users of our libraries would not be broken.

Since JS lacked an existing tracing story and it was originally believed OT would reach 1.0 sometime around September 2019, we originally decided to not abstract away tracing ourselves and instead directly exposed constructs from OT to our customers. This meant that we re-exported `SpanOptions` as public API surface to enable manual span propagation.

As we here in 2020 know, that things didn't happen as planned (with good reason, to be fair), and there ended up being numerous changes to the OT tracing APIs. This put us in a tricky place, since our public interfaces that were supposed to never break had to change in incompatible ways.

We ended up restricting our public surface down to a compatible set of interfaces rather than directly re-exporting shapes from OT: https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/core/core-tracing/src/interfaces.ts

This seemed like it would be good enough to prevent future breaks as the tracing API had largely stabilized. However, that again might have been being too optimistic.

Rather than dropping support for OT, I'm proposing we do the work we avoided doing last year, and invent some abstraction that prevents our customers from having to worry about OT changes, but still allows them to plug into OT if they so desire. It would eliminate any runtime dependency from the Azure SDKs on OT and simplify our bundling story.

Reply:
Cool, thanks for all the info. I reverted the changes and updated the PR. Now, I just moved the files into the API package. Should I rename the function to `isSpanContextValid()`, wasn't sure if the spec was strict about it just being `isValid`. That's how it was before, but `api.isValid()` feels weird.

Reply:
I think it would be better to be `api.trace.isSpanContextValid` @obecny @mayurkale22 wdyt?

Naming does not need to be strict. The API spec allows languages to deviate slightly in naming where it makes sense.

Reply:
This is what spec says https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#isvalid - the change was done 14 days ago. 

For me it looks like the `isValid` should be implemented and placed the same way we have `IsRemote`. Although I can agree that having `isValid` is redundant as it can be calculated from traceId and spanId - I have added my comment [here](https://github.com/open-telemetry/opentelemetry-specification/pull/771#issuecomment-679416550). But if the spec insists on implementing the api "isValid" then I'm afraid we can't do much here and we have to follow the spec. This is rather a spec issue and we should not have incompatibility in this place. I would wait and see what will be the response from spec.

Reply:
lets have what @dyladan already mentioned `api.trace.isSpanContextValid` and then we can get rid of this from context

Reply:
>For me it looks like the isValid should be implemented and placed the same way we have IsRemote

I think this would be fine. If the API package `SpanContext` continues to be an Interface, but just adds an `isValid`, then the "real" implemention is in the SDK, just like it is now. I don't see why this change needs to be such a big change

Reply:
@dyladan so when I started, I just wanted to have an `isValid` in the interface with a default implementation, but TypeScript doesn't allow default methods in an interface. Are you saying to just add ```isValid: () => boolean``` to the spanContext? I'm not sure how that helps unless I'm misunderstanding something.

Reply:
Yes. Then it would be up to the SDK implementation to implement it, just like `isRemote`. @obecny @xirzec how does that sound to you?

@srjames90 hold off on impl until we have some consensus.

Reply:
Fasho, I see what you're saying. Then B3propagator and HttpTraceContext will just add the implementation to `setExtractedSpanContext` like `isRemote`.

Fashi, will hold off.

Reply:
My proposed solution would be to have SpanContext be a class with an implementation of `isValid`, but have that class live in the SDK. The API would still just be an interface, which would now contain `isValid: () => boolean`. @obecny @xirzec?

Reply:
Aw, okay, similar to what I had originally, except without all of the other changes I had where I changed all of the object-literal instantiations?

Reply:
If I understand you correctly, you are suggesting that the concrete implementation of `SpanContext` created by the package possesses an `isValid`, but the public interface doesn't have such a method? 

That seems very reasonable to me, as long as you treat passed in contexts (such as through `SpanOptions` as the interface and not the class. We do this same trick in the Azure SDKs.

Reply:
`isValid` would still be a part of the public interface. But at this point, I think it is easier and less disruptive to just make a function `isSpanContextValid(spanContext: SpanContext): boolean` in the API package, which is what I suggested here: https://github.com/open-telemetry/opentelemetry-js/pull/1447#issuecomment-680183111